src/services/codefixes/fixUnusedIdentifier.ts

```diff
@@ -118,25 +116,34 @@ namespace ts.codefix {
                         }
 
                     case SyntaxKind.NamespaceImport:
-                        const namespaceImport = <NamespaceImport>parent;
-                        if (namespaceImport.name === identifier && !(<ImportClause>namespaceImport.parent).name) {
-                            const importDecl = getAncestor(namespaceImport, SyntaxKind.ImportDeclaration);
-                            return [deleteNode(importDecl)];
-                        }
-                        else {
-                            const previousToken = getTokenAtPosition(sourceFile, namespaceImport.pos - 1, /*includeJsDocComment*/ false);
-                            if (previousToken && previousToken.kind === SyntaxKind.CommaToken) {
-                                const startPosition = textChanges.getAdjustedStartPosition(sourceFile, previousToken, {}, textChanges.Position.FullStart);
-                                return [deleteRange({ pos: startPosition, end: namespaceImport.end })];
-                            }
-                            return [deleteRange(namespaceImport)];
-                        }
+                        return deleteNamedImportBinding(<NamespaceImport>parent);
 
                     default:
                         return [deleteDefault()];
                 }
             }
 
+            function deleteNamedImportBinding(namedBindings: NamedImportBindings): CodeAction[] | undefined {
+                if ((<ImportClause>namedBindings.parent).name) {
+                    // Delete named imports while preserving the default import
+                    // import d|, * as ns| from './file'
+                    // import d|, { a }| from './file'
+                    const previousToken = getTokenAtPosition(sourceFile, namedBindings.pos - 1, /*includeJsDocComment*/ false);
+                    if (previousToken && previousToken.kind === SyntaxKind.CommaToken) {
+                        const startPosition = textChanges.getAdjustedStartPosition(sourceFile, previousToken, {}, textChanges.Position.FullStart);
+                        return [deleteRange({ pos: startPosition, end: namedBindings.end })];
+                    }
+                    return undefined;

```

Comment:
I had my questions about this too. 

My understanding is that ultimately this undefined will be returned from ```getCodeActions``` which is treated by the ```codeFixProvider``` to mean no code fixes are available. So I'm pretty confident returning ```undefined``` is valid and will be handled gracefully.

However I am confused about how to handle seemingly impossible cases (like this one). I don't expect to ever reach this branch and I don't have any test coverage for it. I returned ```undefined``` here to avoid deleting code in an unexpected case. But the previous version of the code would have returned ```return [deleteRange(namespaceImport)];``` and I'm not sure why. Asking @aozgaa ...

I'm not sure about ```[undefined]```  -- I have a feeling VS would not take that too well.